File: src/types/layout.ts
/**
 * How a single field is rendered inside a layout section column.
 *  - `field`: standard label + input
 *  - `image-{size}`: image preview at a preset size (image fields only)
 *  - `table-row`: row in a key:value table; consecutive `table-row`
 *    items in the same column collapse into one table.
 */
export type LayoutDisplay =
  | "field"
  | "image-small"
  | "image-medium"
  | "image-large"
  | "table-row";

export type LayoutItem = {
  /** Stable id; backend assigns one when missing. */
  id?: string;
  fieldKey: string;
  columnIndex: number;
  display: LayoutDisplay;
};

export type LayoutSection = {
  id?: string;
  title?: string;
  /** 1, 2 or 3. */
  columns: number;
  items: LayoutItem[];
};

File: src-tauri/src/domain/layout.rs
use serde::{Deserialize, Serialize};

/// How a single field is rendered inside a section column.
/// - `Field`: standard label + input (the default).
/// - `ImageSmall|Medium|Large`: image preview at a preset size; only
///   meaningful when the underlying field is of image type.
/// - `TableRow`: row in a key:value table. Consecutive `TableRow` items
///   inside the same column collapse into a single rendered table.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum LayoutDisplay {
    Field,
    ImageSmall,
    ImageMedium,
    ImageLarge,
    TableRow,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LayoutItem {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    pub field_key: String,
    pub column_index: i64,
    pub display: LayoutDisplay,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LayoutSection {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    pub columns: i64,
    #[serde(default)]
    pub items: Vec<LayoutItem>,
}

pub fn display_str(d: &LayoutDisplay) -> &'static str {
    match d {
        LayoutDisplay::Field => "field",
        LayoutDisplay::ImageSmall => "image-small",
        LayoutDisplay::ImageMedium => "image-medium",
        LayoutDisplay::ImageLarge => "image-large",
        LayoutDisplay::TableRow => "table-row",
    }
}

pub fn parse_display(s: &str) -> LayoutDisplay {
    match s {
        "image-small" => LayoutDisplay::ImageSmall,
        "image-medium" => LayoutDisplay::ImageMedium,
        "image-large" => LayoutDisplay::ImageLarge,
        "table-row" => LayoutDisplay::TableRow,
        _ => LayoutDisplay::Field,
    }
}
